### PR TITLE
Feature schema validation

### LINKED ISSUE #88: Runtime type validation
All input elements ingest strings, and at runtime all typescript type information is no longer available, so strings can easily end up in the model. 

https://github.com/thegrims/UsTaxes/pull/82/files#diff-464829cf1923068e6ce603cab38ebb785416301be67889ab3e2f0da74e04a432R25

We need to handle all these type errors just at the interface between the user and the action creators, so that any further analysis depending on the global data model can assume the types are correct.

Possible library:
https://github.com/vega/ts-json-schema-generator

### PR DESCRIPTION
**Pull request by zakpatterson**
Fixes #88 

Add typescript json schema generator and validation against schema to action creators.

`redux/data.ts` is the single source of type information for model fields, and `validation.json` is autogenerated to validate types at runtime.

Todo: 
* :heavy_check_mark:  `validation.json` is generated and should not be modified, so it probably doesn't need to be committed either.  So we need to generate it at compile time, or at pull time.
* :heavy_check_mark:  track down more runtime errors 
* :x:  Should be able to use the same validator to validate rehydrated persisted data
    * Will follow up on this later

### CODE REVIEW

**Comment by zakpatterson** [comment]
@thegrims I'm not sure why netlify is failing, the build logs don't help. There's a postinstall script that will generate a validation schema json file. This file has to be generated before the project will compile. Let me know what you think.

**Comment by thegrims** [comment]
@zakpatterson I noticed in the commit where the build started failing 

> Add husky merge hook to refresh validation json

Most of package-lock.json is removed. If that is restored I think the build should run successfully

File: src/redux/actions.ts [hunk 1]
```diff
@@ -50,22 +52,25 @@ function signalAction<T extends ActionName> (t: T): Save<T, {}> {
   }
 }
 
-function makeAction<A extends Object, T extends ActionName> (t: T): ((formData: A) => Save<T, A>) {
+function makeAction<A extends Object, T extends ActionName> (t: T, validate: ValidateFunction<A>): ((formData: A) => Save<T, A>) {
```
Comment (thegrims) [inline src/redux/actions.ts:55]: Can makeAction and makeAction2 be renamed to be more descriptive

Comment (zakpatterson) [reply src/redux/actions.ts]: better?

**Comment by zakpatterson** [comment]
> @zakpatterson I noticed in the commit where the build started failing
> 
> > Add husky merge hook to refresh validation json
> 
> Most of package-lock.json is removed. If that is restored I think the build should run successfully

Ah, you are correct, doing a fresh npm install and pushing that lock file worked!

**Comment by zakpatterson** [comment]
Ah, there is another build error now: 

```
6:50:03 PM: ​
6:50:03 PM: $ CI=false npm run lint && npm run build
6:50:03 PM: > opentaxes@0.1.0 lint /opt/build/repo
6:50:03 PM: > eslint ./src --ext .ts --ext .js --ext .tsx --ext .jsx
6:50:20 PM: > opentaxes@0.1.0 build /opt/build/repo
6:50:20 PM: > react-scripts build
6:50:22 PM: Creating an optimized production build...
6:50:27 PM: Failed to compile.
6:50:27 PM: 
6:50:27 PM: ./src/redux/validate.ts
6:50:27 PM: Cannot find file './validation.json' in './src/redux'.
``` 

That .json file is generated by running `npm install` now. Does netlify not run `npm install` on deploy?

**Comment by thegrims** [comment]
> Ah, there is another build error now:
> 
> ```
> 6:50:03 PM: ​
> 6:50:03 PM: $ CI=false npm run lint && npm run build
> 6:50:03 PM: > opentaxes@0.1.0 lint /opt/build/repo
> 6:50:03 PM: > eslint ./src --ext .ts --ext .js --ext .tsx --ext .jsx
> 6:50:20 PM: > opentaxes@0.1.0 build /opt/build/repo
> 6:50:20 PM: > react-scripts build
> 6:50:22 PM: Creating an optimized production build...
> 6:50:27 PM: Failed to compile.
> 6:50:27 PM: 
> 6:50:27 PM: ./src/redux/validate.ts
> 6:50:27 PM: Cannot find file './validation.json' in './src/redux'.
> ```
> 
> That .json file is generated by running `npm install` now. Does netlify not run `npm install` on deploy?

I don't think it's run every build, they have some sort of npm install dependency caching going on to speed up the build process

**Comment by zakpatterson** [comment]
Ah ok, I think I have it now then.

File: src/redux/actions.ts (hunk 1 above)
Comment (thegrims) [reply src/redux/actions.ts]: Looks great!

**Pull request merged**